test/common/router/router_test.cc

```diff
@@ -2117,5 +2139,36 @@ TEST_F(RouterTestChildSpan, ResetFlow) {
   encoder.stream_.resetStream(Http::StreamResetReason::RemoteReset);
 }
 
+class RouterTestStartTimestampHeader : public RouterTestBase {
+public:
+  RouterTestStartTimestampHeader() : RouterTestBase(false, "X-Request-Start") {}
+};
+
+TEST_F(RouterTestStartTimestampHeader, WithStartTimestampHeader) {
+  absl::optional<Http::Protocol> downstream_protocol{Http::Protocol::Http2};
+  EXPECT_CALL(*cm_.thread_local_cluster_.cluster_.info_, features())
+      .WillOnce(Return(Upstream::ClusterInfo::Features::USE_DOWNSTREAM_PROTOCOL));
+  EXPECT_CALL(callbacks_.request_info_, protocol()).WillOnce(ReturnPointee(&downstream_protocol));
+
+  EXPECT_CALL(cm_, httpConnPoolForCluster(_, _, Http::Protocol::Http2, _));
+  EXPECT_CALL(cm_.conn_pool_, newStream(_, _)).WillOnce(Return(&cancellable_));
+  expectResponseTimerCreate();
+
+  Http::TestHeaderMapImpl headers;
+  HttpTestUtility::addDefaultHeaders(headers);
```

Comment:
test that the `x-request-start` header is not present before the call to decodeHeaders

Reply:
Added in 3110ce0.